### PR TITLE
feat: opening project file via file association

### PR DESCRIPTION
**Pull request by Yukinyaa**
- when only one command line option is given, assume it as project file and try to open it
 - implementation of https://github.com/have-fun-was-taken/yafc-ce/issues/149

### CODE REVIEW

**Comment by Yukinyaa** [comment]
Will open after actually tested

**Comment by have-fun-was-taken** [comment]
> Will open after actually tested

You can make this PR a draft instead of closing it. I think thematically it fits more.

**Comment by Yukinyaa** [comment]
Okey dokey

**Comment by have-fun-was-taken** [comment]
> Merge branch 'master' into master

When the PR is ready for the merge, you will need to rebase the branch with `git rebase -i` to squash merge commits.

File: Yafc/Utils/CommandLineParser.cs [hunk 1]
```diff
@@ -15,6 +16,17 @@ public static class CommandLineParser {
                 return options;
             }
 
+            if (args.Length == 1 && !args[0].StartsWith("--")) {
+                string path = Path.GetFullPath(args[0]);
+                ProjectDefinition? foundproject = Preferences.Instance.recentProjects.Where(
```
Comment (have-fun-was-taken) [inline Yafc/Utils/CommandLineParser.cs:21]:
```suggestion
                ProjectDefinition? foundproject = Preferences.Instance.recentProjects.Where(x => (x.path != null) &&
```

**Comment by have-fun-was-taken** [comment]
In [cbad734](https://github.com/have-fun-was-taken/yafc-ce/pull/152/commits/cbad734be6d7e037ac53f77e6ba80301957e893f), what does `- perf: comparing` mean in the commit message?

**Comment by Yukinyaa** [comment]
* rebase done, thought I had to delete and re-fork to make history clear 👍 
 * 

> In [cbad734](https://github.com/have-fun-was-taken/yafc-ce/pull/152/commits/cbad734be6d7e037ac53f77e6ba80301957e893f), what does `- perf: comparing` mean in the commit message?
In [47d1f6e](https://github.com/have-fun-was-taken/yafc-ce/pull/152/commits/47d1f6e287da97d0f334ed6ad53f588c7ca32b23)(changed due to rebase),  null check is done before `Path.GetFullPath` is called. Should have been more concise, sry😢

**Comment by have-fun-was-taken** [comment]
`implimentation` -> `implementation`

To check the grammar, I usually paste the message in google docs to see if it detects anything.

**Comment by have-fun-was-taken** [comment]
> feat: opening project file via file association
> fix: actually test opening project file
> fix: remove log and redundent code

I have a suspicion that all three commits can be squashed into one because the first one adds logic, when the other two fix that logic to do what was expected. What do you think?

**Comment by Yukinyaa** [comment]
> I have a suspicion that all three commits can be squashed into one because the first one adds logic, when the other two fix that logic to do what was expected. What do you think?

I thought you were talking about squashed merge. I'm OK with squashing all commits in this case. Will do. And also will check for obvious spelling/grammar mistakes. :)

**Comment by have-fun-was-taken** [comment]
> I thought you were talking about squashed merge.  

I'll quote my initial message about that:
> [...] to squash merge commits and other commits that were superseded

By `other commits that were superseded` I meant the commits that fix the new functionality that the PR was aimed to bring. For example, if you make a change that doesn't initially work so you have to fix it _in the code that you introduced_, then you squash the fixes before the PR, unless the fixes are important by themselves for the understanding of the code.

**Comment by Yukinyaa** [comment]
Thank you for the clarification.

I did not realize that was the intended meaning in the Contributor's Guide. I will definitely comply with the guidelines from now on. I'm still getting the hang of contributing to open source projects, and I appreciate this learning opportunity.

**Comment by have-fun-was-taken** [comment]
> I did not realize that was the intended meaning in the Contributor's Guide.

It's fine. This part was not in the Contributor's Guide because I want to keep the guide concise. The think that could be misinterpreted in our case is the following line from it:
> Please separate refactoring and the change of behavior into different commits, so it is easier to review the PR.

This part is about changing the already-existing code. In your current case, if you refactor the new code that the PR brings, then you can squash the refactors.

**Review by veger — COMMENTED** [review]
Code looks good to me, I do have two points:
1. Need to update the command line help (see line https://github.com/have-fun-was-taken/yafc-ce/pull/152/files#diff-b8c38254848464f7da94dda8a1cfe8cb8d22e19b8b0c78b32d630061ae21c1b1R86)
2. Add an entry to the ~changelog.md~ changelog.txt

**Comment by have-fun-was-taken** [comment]
> changelog.md

you mean `changelog.txt`?

**Comment by veger** [comment]
oops, yes

**Comment by have-fun-was-taken** [comment]
@Yukinyaa please add a short summary of the change to the changelog.

File: Yafc/Utils/CommandLineParser.cs [hunk 2]
```diff
@@ -15,6 +16,15 @@ public static class CommandLineParser {
                 return options;
             }
 
+            if (args.Length == 1 && !args[0].StartsWith("--")) {
+                string path = Path.GetFullPath(args[0]);
+                ProjectDefinition? foundProject = Preferences.Instance.recentProjects.FirstOrDefault(
+                        x => x.path != null &&
+                             string.Equals(path, Path.GetFullPath(x.path), StringComparison.InvariantCultureIgnoreCase)
+                    );
+                return foundProject;
```
Comment (Dorus) [inline Yafc/Utils/CommandLineParser.cs:25]: This is a nice way to work with files we've opened before, but what about a fallback for new files? Possibly fall back to the most recent used settings or something? Might want to set path but not skip the welcome screen.

**Comment by have-fun-was-taken** [comment]
Since there are things to change in this PR, and they have not been addressed for a while, I'm converting it to a Draft.

**Comment by shpaass** [comment]
Since there has been no edits to this PR for months and since the fork of @Yukinyaa does not allow edits, I've transferred the PR to [a repo branch](https://github.com/shpaass/yafc-ce/tree/one-arg-welcome-skip) in order to continue the development.

File: Yafc/Utils/CommandLineParser.cs [hunk 3]
```diff
@@ -15,6 +16,15 @@ public static class CommandLineParser {
                 return options;
             }
 
+            if (args.Length == 1 && !args[0].StartsWith("--")) {
+                string path = Path.GetFullPath(args[0]);
+                ProjectDefinition? foundProject = Preferences.Instance.recentProjects.FirstOrDefault(
+                        x => x.path != null &&
```
[the remark this answers is not in the document]
Comment (shpaass) [reply Yafc/Utils/CommandLineParser.cs:22]: Addressed in https://github.com/shpaass/yafc-ce/pull/244/commits/8ddb39fcb4dfa24191649385ee4237b9313b93bd

File: Yafc/Utils/CommandLineParser.cs (hunk 2 above)
Comment (shpaass) [reply Yafc/Utils/CommandLineParser.cs:25]: Here's how I see the situation:

The goal of this feature is to skip the Welcome Screen when a Yafc project is opened with Windows file association. In other words, when a project is opened with something like a double-click on the Yafc project-file.

This functionality is not intended for a seamless creation of a new project. If a player uses this functionality to create a new project, we can just default to the Welcome Screen where they will create it as usual. 

Making this feature work for the creation of a new project may bring more bugs due to edge cases.

File: Yafc/Utils/CommandLineParser.cs
[the remark this answers is not in the document]
Comment (shpaass) [reply Yafc/Utils/CommandLineParser.cs:1]: Addressed in https://github.com/shpaass/yafc-ce/pull/244/commits/3189ecc353a8af68eab4b75e5a1b6ed457228112

File: Yafc/Utils/CommandLineParser.cs (hunk 2 above)
Comment (Dorus) [reply Yafc/Utils/CommandLineParser.cs:25]: In my case it was a copied or moved file. But yes, my preferabled solution would too be to open the welcome screen, optionally prefilled with something sane like the most recent paths.

Comment (shpaass) [reply Yafc/Utils/CommandLineParser.cs:25]: Hmmm, now that I think about it, you're right. The frustration of setting up the full path is much bigger than the frustration of crashing during the load if one had wrong pre-fills. I'll note it in the new PR https://github.com/shpaass/yafc-ce/pull/244. Let's continue our discussion there.

**Pull request closed without merging**